hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/RDBSnapshotProvider.java

```diff
@@ -0,0 +1,233 @@
+  private FaultInjector injector;
+
+  public RDBSnapshotProvider(File snapshotDir, String dbName) {
+    this.snapshotDir = snapshotDir;
+    this.candidateDir = new File(snapshotDir, dbName + SNAPSHOT_CANDIDATE_DIR);
+    this.dbName = dbName;
+    this.injector = null;
+    this.lastLeaderRef = new AtomicReference<>(null);
+    this.numDownloaded = new AtomicLong();
+    init();
+  }
+
+  /**
+   * Initialize or reinitialize the RDB snapshot provider.
+   */
+  public synchronized void init() {
+    // check parent snapshot dir
+    if (!snapshotDir.exists()) {
+      HddsUtils.createDir(snapshotDir.toString());
+    }
+
+    LOG.info("Cleaning up the candidate dir: {}", candidateDir);
+    // cleanup candidate dir
+    if (candidateDir.exists()) {
+      FileUtil.fullyDeleteContents(candidateDir);
+    } else {
+      // create candidate dir
+      HddsUtils.createDir(candidateDir.toString());
+    }
+
+    // reset leader info
+    lastLeaderRef.set(null);
+  }
+
+  /**
+   * Download the latest DB snapshot(checkpoint) from the Leader.
+   *
+   * @param leaderNodeID the ID of leader node
+   * @return {@link DBCheckpoint}
+   * @throws IOException
+   */
+  public DBCheckpoint downloadDBSnapshotFromLeader(String leaderNodeID)
+      throws IOException {
+    LOG.info("Prepare to download the snapshot from leader OM {} and " +
+        "reloading state from the snapshot.", leaderNodeID);
+    checkLeaderConsistent(leaderNodeID);
+
+    String snapshotFileName = getSnapshotFileName(leaderNodeID);
+    File targetFile = new File(snapshotDir, snapshotFileName);
+    downloadSnapshot(leaderNodeID, targetFile);
+    LOG.info("Successfully download the latest snapshot {} from leader OM: {}",
+        targetFile, leaderNodeID);
+
+    RocksDBCheckpoint checkpoint = getCheckpointFromSnapshotFile(targetFile,
+        candidateDir, true);
+    LOG.info("Successfully untar the downloaded snapshot {} at {}.", targetFile,
+        checkpoint.getCheckpointLocation());
+
+    numDownloaded.incrementAndGet();
+    injectPause();
+    return checkpoint;
+  }
+
+  /**
+   * Clean up the candidate DB for the following reason:
+   * 1. If leader changes when installing incremental snapshot
+   *    Notice: here prevents downloading the error IC from the new leader,
+   *    instead, will ask for a full snapshot directly
+   * 2. Ready to download the full snapshot
+   *
+   * @param currentLeader the ID of leader node
+   */
+  private void checkLeaderConsistent(String currentLeader) {
+    String lastLeader = lastLeaderRef.get();
+    if (lastLeader != null) {
+      if (!lastLeader.equals(currentLeader)) {
+        LOG.info("Last leader for install snapshot is {}, but current leader " +
+            "is {}. ", lastLeader, currentLeader);
+        init();
+        lastLeaderRef.set(currentLeader);
+      }
+      return;
+    }
+
+    List<String> files = HAUtils.getExistingSstFiles(candidateDir);
+    if (!files.isEmpty()) {
```

Comment:
In what case that this condition will stand?

Reply:
This is a prospective case. For example, this follower just gets restarted and lost the in-memory info about whose DB this candidate dir stores(lastLeaderRef).  
Since how long the recovery time is not certain, even if the leader is not changed during the recovery time, the old candidate DB might be outdated or modified. To simplify the solution, here just add the warn log and reinitialize the candidate DB directory.